lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs

```diff
@@ -133,37 +116,22 @@ public byte[] Call(byte[] memento)
                 _dataReducer.Send(result);
             }
 
-            Interlocked.Exchange(ref _isTaskStopped, 1);
+            _taskCloseHandler.SetTaskStopped();
 
-            if (Interlocked.Read(ref _shouldCloseTask) == 1)
+            if (_taskCloseHandler.ShouldCloseTask())
             {
-                _waitToCloseEvent.Set();
+                _taskCloseHandler.Signal();
```

Comment:
Close is asynchronous. It's possible that the close Event handler comes in a bit later and in that case it will hang.